### ISSUE TITLE
Loading state during navigation (not prefetching)

### ISSUE THREAD

**Issue by nfiacco** [vercel/next.js#58261]
Would love a way to show instant feedback to the user when they navigate between routes that doesn't involve prefetching.

I don't think the point about 100KB loading states is a valid concern. Loading states are often just a spinner. If anything, the default approach should optimize for basic loading states and let people configure something special if they have a special 100KB loading state.

Let me know your thoughts, definitely want to understand how the Next team is thinking about this.

> I want to clarify how loading UI works in Next.js.
> 
> Routing in the App Router is "server centric". We need a network roundtrip to the server when doing route transitions. The client doesn't know what loading state to render, until it communicates with the server.
> 
> When we prefetch, we can retrieve both the route to navigate to _and_ the loading UI. Prefetching is what makes navigations feel snappy. If you prefetch the loading UI which does not have data fetching, then you can "instantly" show this loading state, followed by the dynamic content on the page being streamed in after. This is enabled through React Suspense and streaming server-rendering as part of the App Router.
> 
> We don't believe an alternate solution where you include all of the routing variants on the client is feasible or the best solution for Next.js. For example, consider a loading state which is `100kb` of HTML / possible routes. Should that be included in the initial JavaScript bundle, blocking interactivity for *everything* on the page until it's downloaded? Or, is it better to *not* block interactivity, and load it asynchronously before the next click? We still believe the current model is correct.
> 
> There's still opportunities for us to improve the experience of transitioning between loading states. For example, we can optimize further the de-duplication of data. Thanks for the reproductions here, as they'll help us dig in further and find places to optimize.
> 
> _Originally posted by @leerob in https://github.com/vercel/next.js/issues/43548#issuecomment-1758745511_
>

**Comment by philipheinser** [comment]
@nfiacco @leerob  I agree its a issue and feels like a bug when you use a nexts page. They write in docs:

> An instant loading state is fallback UI that is shown immediately upon navigation.

The comment you postet they explain why they cannot show an instant loading state. For us a really simple loading ui that does not need to be `100kb` at all would love the issue for us. Right now the user get no feedback at all when clicking a link on slow connection its feels super buggy.

**Comment by vankatadam** [comment]
@nfiacco @leerob @timneutkens Same here. Another Problem is, that the loading.jsx/tsx is blocked till generateMetadata is ready. This is not a problem in HelloWorld examples, where only static metadata is used, but in dynamic routes you have to firstly fetch your data for your metadata, which can take some time. It's a terrible UX, if the user has to wait for that, till he gets any Loading Feedback on a simple Next Link. Its like wait for 2 Seconds of nothing happening, than for 2 Seconds there is the loading.tsx and than the page loads. Thats almost worse than not having anything and worse than even the standard behavior of a simple <a> where the user atleast sees, that something is happening in his browser bar. It would be great, if we had some control of when we can load our loading.tsx and decide on our own, if it is too big to be prefetched, or atleast be fetched as a first thing that the server does, before generateMetadata. Thanks in advance for your answer, or explanation of the decision behind this implementation

**Comment by purplecandy** [comment]
@leerob any updates on this? 
Here's a short preview of how it breaks the experience and feels like things aren't responding 
![Screenity video - Jan 14, 2024 (1)](https://github.com/vercel/next.js/assets/17525936/e7a0c458-9c91-4386-bf66-2babdb6e14b9)

Compared to how Github does it with a small loading progress indicator makes all the difference. 
![Screenity video - Jan 14, 2024](https://github.com/vercel/next.js/assets/17525936/e02d4cb8-9b99-4533-a5ae-131151582519)

**Comment by nick-barth** [comment]
This comment is not meant to be a threat or an ultimatum, but if instant navigation is not fixed within a month I will be forced to refactor my entire frontend to Remix or Sveltekit because this is such bad user experience, our app appears broken, and EVERYONE notices.

**Comment by dragos99** [comment]
+1 for this. NextJS is a bad user experience until this is addressed.

**Comment by AiAutomationAgency** [comment]
yes, this is a bad user experience, hhhh, it's really funny, "Nextjs" came to solve and improve the user experience and SEO, It fixed the SEO and made the user experience worse.

**Comment by tonyjara** [comment]
This is honestly a big pain point and I wish it would get more attention. I've had to do super lightweight skeletons pages to mitigate this problem. It would be so nice to get a simple state update whenever the next page is getting loaded so we can show stuff like a spinner in the button pressed.

**Comment by selcukcihan** [comment]
Not entirely sure if the pain point described here is the same as the problem I'm trying to solve, but it definitely sounds similar. Google search brought me here. My problem is, I have a single page and I'm using query search parameters to display content for different dates, for instance

localhost:3000 shows todays content
localhost:3000?date=2024-06-03 shows yesterday's etc.

I also have a drop down menu that displays available date selections, which are `Link`s of `?date={selectedDate}`. My problem is, I want to display a loading spinner when one of those links are clicked, until the page is fully rendered. How can I accomplish this? I've tried the loading.tsx but I think it's not working as it's the same page. For the forms on my page, I'm using `useFormStatus` to show a loading UI until the app receives a response, I want something similar here. Any ideas?

**Comment by patateskafa** [comment]
I've been switching to the App Router and learning its tricks for the last 2-3 weeks, and I never questioned whether this could be an actual issue. I kept reading 'An instant loading state is fallback UI that is shown immediately upon navigation' in the docs, tried explicit Suspend wrappers, etc., and couldn't solve it. Impostor syndrome kicked in many times, making me wonder, 'What am I doing wrong here?' until I read the description of this issue (sadly, it was not among the top results in a Google search, maybe because it's from November 2023). Wow, I'm indeed shocked. I think at least the current behavior and this particular shortcoming should be mentioned in the docs in CAPITAL LETTERS. instant navigation feedback is one of the simplest yet most impactful details regarding the UX of modern web apps, it's a no-brainer right? anyway, just my 2 cents to emphasize the importance of this. I appreciate all the hard work that goes into Next, so thank you to the hardworking fellas at Vercel (not to the company itself, to the workers only :') )

**Comment by patateskafa** [comment]
hi Lee. there could be many workarounds indeed but, I think most of the people here (+me) are asking for a "native Next" solution regarding the loading state navigation behavior, making it instant (like, "client instant", not "server instant") (after I read this issue I discovered the existence of others i.e. https://github.com/vercel/next.js/issues/43548 )

but even if that's not on the roadmap I think there should at least be a disclaimer in the docs here: https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming  `An instant loading state is fallback UI that is shown immediately upon navigation.` is misleading as it's not 100% accurate for non-prefetched scenarios (and sometimes for prefetching as well, if the request fails etc.) and it can save newcomers some time.

**Comment by dragos99** [comment]
I believe partial prerendering is the native solution to this @patateskafa. Still experimental, but perhaps coming soon 🙏🏼

**Comment by philipheinser** [comment]
> It sounds like there are few product experiences here where you are all wanting to _optimistically_ show an update, immediately. Is that correct? 
> 
> 
> 
> For example, looking at this [comment](https://github.com/vercel/next.js/issues/58261#issuecomment-1890937827). You could `useOptimistic` here and have `isPending` to showing a loading animation at the component level. Or you could do a similar "global progress bar" pattern with something like [`react-transition-progress`](https://github.com/vercel/react-transition-progress).

We should show the user progress right when he clicks. Right now we implementet global progress bar as a workaround but it is not good to do it in user land. We dont know if the site we navigate to is already prefetched or not so the bar always apears even if not needed making the site feel slower. It should be implemented on router level. Even in plain html you have a progress bar from the browser when a new page loads. The default nextjs expierence is worse.

**Comment by khanakia** [comment]
Any update i have same issue to user it feels like nothing happening when he clicks the Link

**Comment by jvandenaardweg** [comment]
So, i guess there's some timeout happening under the hood? Whenever I artificially delay a server component for 2 seconds, it does not show my loading.tsx . However, if i increase the delay, to say 3 seconds, it finally shows it.

I have defined a loading.tsx in my route, so why is this even happening? Just show the loading.tsx

**Comment by aldenquimby** [comment]
@jvandenaardweg my team is doing the same thing @philipheinser mentioned - all links use a client-side wrapper around next/link that shows a global loading indicator immediately onClick, and hides it when the route transition is complete (with a min of 1s so the loader doesn't flash)

It's unfortunate to lose the benefits of the native Link, and to show a 1s loader unnecessarily when a route is prefetched, but overall it creates and way better UX

**Comment by charanjeetroshan** [comment]
After struggling with the same issue, I came up with a workaround for this problem which is dynamic imports in Next.js. It is the alternative to the `React.lazy()` in the "Next.js world".

Import your component dynamically in your app route and render it.

```
'use client'
 
import dynamic from 'next/dynamic'
 
const WithCustomLoading = dynamic(
  () => import('../components/WithCustomLoading'),
  {
    loading: () => <p>Loading...</p>,
  }
)
 
export default function Page() {
  return (
    <div>
      {/* The loading component will be rendered while  <WithCustomLoading/> is loading */}
      <WithCustomLoading />
    </div>
  )
}
```

Next.js documentation: https://nextjs.org/docs/app/guides/lazy-loading

It is indeed an extra step to take but it works.

**Comment by eskander-taher** [comment]
its already 2 years and they still didn't solve it, what a shame

**Comment by armya-z** [comment]
still no standard solotion?

**Comment by eskander-taher** [comment]
Nope

**Comment by armya-z** [comment]
I ended up implementing a custom Link wrapper since useLinkStatus doesn't work reliably for tracking navigation loading states in the App Router.

```
"use client";

import Link from "next/link";
import { useTransition } from "react";

export default function LinkWithLoading({
  href,
  children,
  className,
}: {
  href: string;
  children: React.ReactNode;
  className?: string;
}) {
  const [isPending, startTransition] = useTransition();

  return (
    <Link
      onClick={() => startTransition(() => {})}
      prefetch={false}
      href={href}
      className={className}
    >
      {children}
      {isPending && (
        <span role="status" aria-label="Loading" className="spinner" />
      )}
    </Link>
  );
}
```
This leverages React’s useTransition to show a loading indicator during client-side navigation. The empty startTransition callback is sufficient to trigger the pending state—no actual work needs to happen inside it.

**Comment by eskander-taher** [comment]
Does this solution show the loading state until the requested page is
fetched, or is just a loading state with a specific duration regardless of
how slow or fast the fetching is happening?

**Comment by armya-z** [comment]
> Does this solution show the loading state until the requested page is
> fetched, or is just a loading state with a specific duration regardless of
> how slow or fast the fetching is happening?
> […](#)

yes it does. after user taps a link, loader will appear till render completes.

**Comment by marcoshernanz** [comment]
Thanks for the feedback. This is now supported by [`useLinkStatus`](https://nextjs.org/docs/app/api-reference/functions/use-link-status), added in Next.js 15.3. It gives a clicked `<Link>` an immediate pending state when prefetching is disabled or has not completed, which covers the case described here. The navigation docs now call out this behavior as well.

Closing as completed.
